### PR TITLE
CE-1543 Fix editcount ET updates

### PR DESCRIPTION
**Pull request by kamilkoterba**
https://wikia-inc.atlassian.net/browse/CE-1543

Previously user record was updated in ExactTarget after each ArticleSaveComplete. Some inconsistencies in editcount user number updates found. Decided to change approach and update numbers once a day based on list of users that edited previous day. (same as it's done for incremental user per wiki daily updates)

This way we gain data consistency and reduce number of updates.
Assume 4k users edited 3 times each. It was causing 12k updates. Now it will be two updates one with 2.5k users and second with remaining 1.5k (chunk limit 2.5k)

@SQLReview 

```
SELECT DISTINCT user_id
     FROM rollup_wiki_user_events
     WHERE time_id > '2015-04-13 02:47:26'
     AND period_id = '1'
     AND user_id <> '0'
```

time_id is always set to be greater than day before

EXPLAIN

```
          id: 1
  select_type: SIMPLE
        table: rollup_wiki_user_events
         type: range
possible_keys: PRIMARY,time_id
          key: PRIMARY
      key_len: 8
          ref: NULL
         rows: 67732
        Extra: Using where; Using index; Using temporary
```

@Wikia/community-engineering

### CODE REVIEW

**Comment by wladekb** [comment]
lgtm from sql perspective

File: extensions/wikia/ExactTargetUpdates/maintenance/ExactTargetUpdateUserGlobalEditCountMaintenance.php [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+		$oWikiaSQL->SELECT()
+			->DISTINCT( 'user_id' )
+			->FROM( 'rollup_wiki_user_events' )
+			->WHERE( 'time_id' )->GREATER_THAN( $sStartDate )
+			->AND_( 'period_id' )->EQUAL_TO( DataMartService::PERIOD_ID_DAILY )
+			->AND_( 'user_id' )->NOT_EQUAL_TO( 0 );
+
+		$aUsersIdsDidEdit = $oWikiaSQL->runLoop( $oStatsDBr, function( &$aUsersIdsDidEdit, $oUsersIdsDidEdit ) {
+			$aUsersIdsDidEdit[] = $oUsersIdsDidEdit->user_id;
+		});
```
Comment (Grunny) [inline extensions/wikia/ExactTargetUpdates/maintenance/ExactTargetUpdateUserGlobalEditCountMaintenance.php]: Should be spacey.

File: extensions/wikia/ExactTargetUpdates/maintenance/ExactTargetUpdateUserGlobalEditCountMaintenance.php [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+		$aUsersData = $this->prepareUsersParams( $aUsersIds );
+		// Add update tasks
+		$this->addUsersUpdateTasks( $aUsersData );
+	}
+
+	/**
+	 * Fetches user ids from statsDB from last period determined by prepareTimeCondition function
+	 * @param DatabaseBase $oStatsDBr
+	 * @param string $sStartDate e.g. 2014-12-31
+	 * @param boolean|ResultWrapper $oUsersListResult
```
Comment (Grunny) [inline extensions/wikia/ExactTargetUpdates/maintenance/ExactTargetUpdateUserGlobalEditCountMaintenance.php]: This parameter doesn't exist?

File: extensions/wikia/ExactTargetUpdates/maintenance/ExactTargetUpdateUserGlobalEditCountMaintenance.php [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+		/* Get and run the task */
+		$task = new ExactTargetUpdateUserTask();
+		$task->call( 'updateFallbackCreateUsers', $aUsersData );
+		$task->queue();
+	}
+
+	private function getLastDayDate() {
+		$oNow = new DateTime();
+		$oNow->sub(new DateInterval('P1D'));
+		$sStartDate = $oNow->format('Y-m-d H:i:s');
```
Comment (Grunny) [inline extensions/wikia/ExactTargetUpdates/maintenance/ExactTargetUpdateUserGlobalEditCountMaintenance.php]: Should be spacey.

File: extensions/wikia/ExactTargetUpdates/tasks/ExactTargetUpdateUserTask.php [hunk 4]
```diff
@@ -15,6 +15,52 @@ class ExactTargetUpdateUserTask extends ExactTargetTask {
...
+
+		/* Validate users data, unset if incomplete */
+		foreach ( $aUsersData as $iArrayKey => $aUserData ) {
+			if ( empty( $aUserData['user_id'] ) ) {
+				$this->error( __METHOD__ . ' user under ' . $iArrayKey . ' index of $aUsersData has no ID' );
+				unset( $aUsersData[$iArrayKey] );
+			}
+
+			elseif ( empty( $aUserData['user_email'] ) ) {
+				$this->error( __METHOD__ . ' user under ' . $iArrayKey . ' index of $aUsersData has no ID' );
```
Comment (Grunny) [inline extensions/wikia/ExactTargetUpdates/tasks/ExactTargetUpdateUserTask.php]: "has no ID" -> "has no email".

File: extensions/wikia/ExactTargetUpdates/tasks/ExactTargetUpdateUserTask.php [hunk 5]
```diff
@@ -15,6 +15,52 @@ class ExactTargetUpdateUserTask extends ExactTargetTask {
...
+	public function updateFallbackCreateUsers( $aUsersData ) {
+
+		/* Validate users data, unset if incomplete */
+		foreach ( $aUsersData as $iArrayKey => $aUserData ) {
+			if ( empty( $aUserData['user_id'] ) ) {
+				$this->error( __METHOD__ . ' user under ' . $iArrayKey . ' index of $aUsersData has no ID' );
+				unset( $aUsersData[$iArrayKey] );
+			}
+
+			elseif ( empty( $aUserData['user_email'] ) ) {
```
Comment (Grunny) [inline extensions/wikia/ExactTargetUpdates/tasks/ExactTargetUpdateUserTask.php]: Move this onto the same line as the closing brace of the `if`.

File: extensions/wikia/ExactTargetUpdates/tasks/ExactTargetUserDataVerificationTask.php [hunk 6]
```diff
@@ -11,22 +11,39 @@ class ExactTargetUserDataVerificationTask extends ExactTargetTask {
 	 * return true if data is equal (except comparing user_touched)
 	 * if data isn't equal throws exception with result diff
 	 */
-	public function verifyUserData( $iUserId ) {
+	public function verifyUsersData( array $aUsersIds ) {
+		$bSummaryResult = true;
 		// Fetch data from ExactTarget
 		$oRetriveveUserTask = $this->getRetrieveUserTask();
```
Comment (Grunny) [inline extensions/wikia/ExactTargetUpdates/tasks/ExactTargetUserDataVerificationTask.php]: `$oRetriveveUserTask` -> `$oRetrieveUserTask`.

File: extensions/wikia/ExactTargetUpdates/tasks/ExactTargetUserTaskHelper.php [hunk 7]
```diff
@@ -349,4 +362,16 @@ public function mergeUsersEditsData( array &$aUsersEditsData, array $aUserEditsD
...
 
+	/**
+	 * Extracts list of users ids from provided $aUsersData
+	 * @param $aUsersData array of users data arrays consistent with structure of ExactTargetUserHooksHelper::prepareUserParams
+	 * @return array
+	 */
+	public function prepareUserIds( $aUsersData ) {
+		foreach( $aUsersData as $aUserData ) {
+			$aUserIds[] = $aUserData['user_id'];
+		}
```
Comment (Grunny) [inline extensions/wikia/ExactTargetUpdates/tasks/ExactTargetUserTaskHelper.php]: If you want to use a new builtin method we got with the upgrade from PHP 5.4 to 5.6, [`array_column`](http://php.net/manual/en/function.array-column.php) will do this in one line: `array_column( $aUserData, 'user_id' );`

File: extensions/wikia/ExactTargetUpdates/tasks/ExactTargetUserDataVerificationTask.php [hunk 8]
```diff
@@ -11,22 +11,39 @@ class ExactTargetUserDataVerificationTask extends ExactTargetTask {
...
-		$oExactTargetUserData = $oRetriveveUserTask->retrieveUserDataById( $iUserId );
-		$this->info( __METHOD__ . ' ExactTarget user data record: ' . json_encode( $oExactTargetUserData ) );
+		$aExactTargetUsersData = $oRetriveveUserTask->retrieveUsersDataByIds( $aUsersIds );
+		$aUsersIdsFlipped = array_flip( $aUsersIds );
+		foreach( $aExactTargetUsersData as $aExactTargetUserData ) {
```
Comment (Grunny) [inline extensions/wikia/ExactTargetUpdates/tasks/ExactTargetUserDataVerificationTask.php]: Space needed between `foreach` and the left parenthesis per conventions.

**Comment by Grunny** [comment]
LGTM.

**Pull request merged**